Support idmap mounts for volumes

rata opened the pull request:
This PR implements support for this runtime-spec change that added idmap mounts support: https://github.com/opencontainers/runtime-spec/pull/1143

We open the idmap source paths and call mount_setattr() in runc PARENT,
as we need privileges in the init userns for that, and then sends the
fds to the child process. For this fd passing we use the same mechanism
used in other parts of thecode, the `_LIBCONTAINER_` env vars.

The mount is finished (unix.MoveMount) from go code, inside the userns,
so we reuse all the prepareBindMount() security checks and the remount
logic for some flags too.

This PR only supports idmap mounts when userns are used AND the mappings
are the same specified for the userns mapping. This limitation is to
simplify the initial implementation, as all our users so far only need
this, and we can avoid sending over netlink the mappings, creating a
userns with this custom mapping, etc. Future PRs will remove this
limitation.

As the idmap case is quite similar to the existing mount sources case we
open with O_PATH, some simple refactors are done to share more code and
to group the slices of fds in go code. To that end, we created the
mountFds struct, and add all the slices of fds there.

This replaces PR #3429, as that PR tries to mount already when we are inside the user namespace. AFAIK, that will never work and therefore this PR tries a completely different way to do the idmap mounts.

This PR is co-authored-by: `Francis Laniel <flaniel@linux.microsoft.com> `

cc @eiffel-fl

Question
------------
One open question I have is I added most validations in `libcontainer/configs/validate/validator.go` but I couldn't help to notice that other parts of the code do some other validations. All examples that I tried hit the validations, but am I missing adding the validations in some other part, that don't use this maybe when runc is called differently? Or adding them there should be enough?


TODO
--------
- [x] Pass idmap mount sources using bootstrap data (now hardcoded an example path)
- [x] Add integration tests
- [x] Modify the mountToRootfs() function to not move_mount and finish the idmap mount
- [x] Update runtime spec to have the new fields for UID/GID mappings
- [x] Address all TODOs I added in this PR

Changelog entry
----------------------

 * Support idmap mounts as specified in the OCI runtime-spec. Currently the mount mappings need to be identical to the mappings used in the user namespace section.
 * Enforce absolute paths for mounts


Closes: #3429 
Closes: #3020

rata commented:
@AkihiroSuda can we add this to the 1.2 milestone?

rata commented:
We have something working now. Will polish and push next week. Still tests and that missing.

rata commented:
Test failures seem unrelated, for example:

```
make[1]: Entering directory '/go/src/github.com/opencontainers/runc'
fatal: detected dubious ownership in repository at '/go/src/github.com/opencontainers/runc'
To add an exception for this directory, call:

	git config --global --add safe.directory /go/src/github.com/opencontainers/runc
go build -trimpath -buildmode=pie -a -tags "seccomp urfave_cli_no_docs netgo osusergo" -ldflags "-X main.gitCommit= -X main.version=1.1.0+dev -linkmode external -extldflags --static-pie -w -s -buildid=" -o runc .
error obtaining VCS status: exit status 128
	Use -buildvcs=false to disable VCS stamping.
make[1]: *** [Makefile:69: static] Error 1
```

or go 1.19 with -race failing due to criu:
```
Run # criu repo
Warning: apt-key output should not be parsed (stdout is not a terminal)
curl: (92) HTTP/2 stream 0 was not closed cleanly: PROTOCOL_ERROR (err 1)
gpg: no valid OpenPGP data found.
Error: Process completed with exit code 2.
```

AkihiroSuda commented on libcontainer/configs/mount_linux.go:
```diff
@@ -29,8 +31,21 @@ type Mount struct {
 	return m.Flags&unix.MS_BIND != 0
 }
+
+func (m *Mount) IsBindOnly() bool {
```
```suggestion
func (m *Mount) IsBindWithoutIDMap() bool {
```

rata replied:
@AkihiroSuda I'm not convinced that is better. It is all pretty cosmetic, so I'll commit your suggestion, but in this matter of personal taste, I'd say I prefer the current name. If tomorrow we have another kind of mount (bind mounts that are XXX too, besides bind mounts that are idmap) then we would want here to be `bindWithoutIDMapWithoutXX()`?

It seems not nice IMHO. While if this is _only_ a bind mount is what we want (not a bindmount that is also idmap, or XXX, etc.).

What do you think?

rata replied:
As we discussed in other comments, I've just removed this and open-code it whenever I need it.  Thanks!

AkihiroSuda commented on libcontainer/configs/validate/validator.go:
```diff
@@ -259,6 +260,36 @@ func cgroupsCheck(config *configs.Config) error {
+		}
+
+		if config.RootlessEUID {
+			return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported in rootless mode, needs CAP_SYS_ADMIN on initial user namespace", m)
```
```suggestion
			return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported in rootless mode, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
```

rata replied:
Fixed, but with s/the runc/runc/

AkihiroSuda commented on libcontainer/container_linux.go:
```diff
@@ -518,57 +518,107 @@ func (c *Container) shouldSendMountSources() bool {
+		return false
+	}
+
+	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in the initial userns.
```
```suggestion
	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in the runc parent's userns
```

rata replied:
@AkihiroSuda it requires it in the initial userns for now (no fs supports idmap mounts inside a userns yet). See the manpage comment: https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/tree/man2/mount_setattr.2#n638

When that is relaxed, maybe we can simplify this and do the mount inside the userns without the nsexec and fd passing (not sure it will be worth as long as we support older kernels). But for now this seems accurate to me.

What do you think?

brauner replied:
You need CAP_SYS_ADMIN in the user namespace the filesystem was mounted in and CAP_SYS_ADMIN in the user namespace you're trying to idmap the mount to. And you need CAP_SYS_ADMIN in the owning user namespace of the mount namespace you're currently located in.

For most filesystems the user ns of the filesystem will be init_user_ns. That'll change with tmpfs being able to support idmapped mounts.

So you're both wrong. ;)

rata replied:
Added a comment with this, thanks!

AkihiroSuda commented:
Please rebase, then the CI should be green

rata commented:
CI still fails with the make release fix, in the same way (git safe dir): https://github.com/opencontainers/runc/actions/runs/4135850345/jobs/7148887122

rata commented:
Uff, and cantos-7 failing with:

```
=== RUN   TestSkipDevicesTrue
    systemd_test.go:136: mkdir /sys/fs/cgroup/hugetlb/system.slice/system-runc_test_pods.slice: no such file or directory
--- FAIL: TestSkipDevicesTrue (0.07s)
```

rata commented:
@kolyshkin I'm not sure what is wrong with that case (I'm not familiar with the cgroup manager thingy). But created an issue for the flaky test

rata commented:
Added integration tests now. There might be some simplifications to do in the bats file (like configure the process to run on the setup, and only override it when we change in a test) but it should be mostly ok.

I'll be afk this next week, but @@eiffel-fl will be online to answer :)

rata commented:
CI takes ages, fails for lot of unrelated reasons so it is hard to debug... I'll leave this to @eiffel-fl to see how to skip the tests in the distros that don't run with idmap support on the fs we are using ;)

brauner commented:
> After a discussion with rata, here are our conclusions:
> 
> 1. The integration test fail for `fedora` because it is run on `tmpfs` which does not support ID map mounts.

This is now supported with kernel v6.3. Linus merged my pull request adding support for this yesterday:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=05e6295f7b5e05f09e369a3eb2882ec5b40fff20

> 2. For `centos`, we do not have any clue as `-EPERM` is returned, possibly from one of these locations:
>    https://elixir.bootlin.com/linux/v5.14/source/fs/namespace.c#L4076
>    https://elixir.bootlin.com/linux/v5.14/source/fs/namei.c#L502
>    https://elixir.bootlin.com/linux/v5.14/source/fs/namespace.c#L3914
>    https://elixir.bootlin.com/linux/v5.14/source/fs/namespace.c#L3873
>    https://elixir.bootlin.com/linux/v5.14/source/fs/namespace.c#L3885
> 
> I will try to add more options until I can boot the `centos` kernel to debug it. Meanwhile, if anyone has any idea why it returns `-EPERM`, feel free to share!

I'm currently still traveling so haven't seen this until now.
The reason why idmapped mounts are disabled by default in CentOS 9 and RHEL is due to a decision made by a single developer.

The developer repeatedly made claims to the effect that idmapped mounts supposedly have issues but never followed up with any proof despite repeatedly being asked to do so. The relevant claims are demonstrably false, and there is no evidence to support them (cf. [1] and [2]). Idmapped mounts are one of the best tested features in all of the vfs, with a dedicated generic and filesystem specific testsuites in xfstests (cf. [3]).

Additionally, the introduction of a dedicated type vfsuid_t/vfsgid_t has implemented extensive type-safety hardening measures to separate uids/gids generated via idmapped mounts from other uids/gids. Starting with kernel v6.3 we even have a dedicated type that separates idmappings attached to a mount from all other idmappings which means they cannot be conflated with each other without the compiler catching the error.

Despite the demonstrable evidence that idmapped mounts are safe and secure, the developer in question has repeatedly made false claims and accusations without providing any proof. They have also sent off-list emails to the Linux Plumbers organizing committee in 2021, repeating their unfounded claims behind the backs of others (cf. [4]). This behavior is unacceptable and does not contribute positively to the community.

While there have been efforts by Giuseppe, Frantisek, and others to get idmapped mounts enabled on RHEL/CentOS 9, it is unfortunate that the actions of one individual have led to this feature being disabled by default. Idmapped mounts should simply be reenabled on RHEL/CentOS 9 and those who make false claims and unfounded accusations should be held accountable for their actions.

```
[1] https://lore.kernel.org/lkml/20210213130042.828076-1-christian.brauner@ubuntu.com/T/#m3a9df31aa183e8797c70bc193040adfd601399ad
    https://lore.kernel.org/lkml/20210213130042.828076-1-christian.brauner@ubuntu.com/T/#m59cdad9630d5a279aeecd0c1f117115144bc15eb
[2] https://lore.kernel.org/lkml/m1r1ifzf8x.fsf@fess.ebiederm.org
    https://lore.kernel.org/lkml/20210510125147.tkgeurcindldiwxg@wittgenstein/
[3] https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git/tree/src/vfs
[4] https://lore.kernel.org/linux-fsdevel/CAHrFyr4AYi_gad7LQ-cJ9Peg=Gt73Sded8k_ZHeRZz=faGzpQA@mail.gmail.com
    https://lore.kernel.org/linux-fsdevel/87pmst4rhy.fsf@disp2133/
    https://lore.kernel.org/linux-fsdevel/CAHrFyr7z82u1QcOo+f72sPr7_L8qUxo1EAK=JeU8xuzAsO9REg@mail.gmail.com/
```

brauner commented:
In short, skip on `-EPERM` on CentOS/RHEL.

AkihiroSuda commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,219 @@
+function setup() {
+	# idmap mounts requires root and kernel >= 5.12
+	requires root
+	requires_kernel 5.12
```
needs `requires_idmap_fs`

rata commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,215 @@
+
+function setup() {
+	# idmap mounts requires root and kernel >= 5.12
+	requires root
+	requires_kernel 5.12
+	requires_idmap_fs /tmp
```
With `require_idmap_fs` I think we can just remove the other requires (unless requires_idmap_fs is very slow, in which case leaving the other requires will speed things up for old kernels or non-root runs)

rata replied:
These doesn't represent better the requirements. Kernel patches can be backported, so testing for 5.12 is not really that good. And we don't really need root, we need CAP^_SYS_ADMIN in several userns.

IMHO testing if we can idmap seems to be exactly what we need for the test

rata replied:
We talked on slack and Francis preferred to leave it as it is now. Otherwise we need to skip the tests in all cases the binary fails, which can easily cause an oversight if we break the binary to always fail.

brauner commented on tests/integration/helpers.bash:
```diff
@@ -620,3 +621,35 @@ function requires_kernel() {
+				# /*
+				#  * So far, there are serious concerns about the safety of idmaps.
+				#  * This mechanism requires further upstream review.
+				#  */
```
> Could you add the URL of this snippet

The nonsensical merge request and commit is:
https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/merge_requests/131
https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/commit/c93c728360fd10a0dc65

brauner replied:
One should also note that it's 2023. So that feature has been in mainline for 3 years now without alleged security issues and with support for at least xfs, ext4, btrfs, overlayfs, f2fs, squashfs, tmpfs, ntfs3, fat, erofs, and more to come. With a dedicated tree as part of my general VFS tree and @sforshee as co-maintainer.

AkihiroSuda commented:
Please sign off the commits
https://github.com/apps/dco

rata commented on tests/integration/helpers.bash:
```diff
@@ -620,3 +621,52 @@ function requires_kernel() {
+		;;
+	*operation\ not\ permitted)
+		if uname -r | grep -q el9; then
+			# centos kernel 5.14.0-200 contains the following snippet in
+			# can_idmap_mount():
```
Christian shared a link, maybe we can put the link here instead of a c&p of the code?

From [this comment](https://github.com/opencontainers/runc/pull/3717#discussion_r1113374333), the code is here:

https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/merge_requests/131
https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/commit/c93c728360fd10a0dc65

brauner reviewed (approved):
Nicely done, thanks!

rata commented:
All test green now and squashed the commits. @AkihiroSuda PTAL, thanks!

rata commented on libcontainer/container_linux.go:
```diff
@@ -2237,6 +2296,28 @@ func (c *Container) bootstrapData(cloneFlags uintptr, nsMaps map[configs.Namespa
+		var mounts []byte
+		for _, m := range c.config.Mounts {
+			if m.IsIDMap() {
+				// TODO: why do we need to duplicate this that is already done in
+				// libcontainer/specconv/spec_linux.go
```
Does anyone know why we need this? I see it was added to avoid json strings to have null characters that will mix with the `byte(0)` we send a few lines below, the possible security issue and all...

But do we need to _duplicate_ this here? I just did it because in the case of `c.shouldSendMountSources()` was added there _too_. But isn't enough with the check done in spec_linux.go? Or if we check for this in the validator?

kolyshkin replied:
This is because in case libcontainer is used by some other software (not runc), the `specconv` is no longer involved.

kolyshkin replied:
@eiffel-fl can you please modify the comment above to explain why this is needed? 

In case libcontainer is used by some other software (not runc), the check we perform in  specconv is not executed, so we also check it here.

rata replied:
Added a comment, thanks!

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
+		mountFds[i] = -1
+		idmapFds[i] = -1
+		switch {
+		case m.IsBindOnly() && c.shouldSendMountSources():
```
w.r.t. abstracting; if we inline some of the logic that's abstracted, this switch effectively becomes;

```go
		case m.IsBind() && len(m.UIDMappings) == 0 && len(m.GIDMappings) == 0 && c.shouldSendMountSources():
		case m.IsBind() && (len(m.UIDMappings) > 0 || len(m.GIDMappings) > 0) && c.shouldSendIdmapSources():
```

At least, it looks like all branches only act if `m.IsBind()`, so we could consider moving that out of the switch (and `continue` early on in the loop (_possibly_ consider moving the shared `cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)` outside of the switch as well.

If we combine some of that, and rewrite this function, it becomes something like this;

```go
func (c *Container) sendSources(cmd *exec.Cmd, messageSockPair filePair) error {
	sendMountSources := c.shouldSendMountSources()
	sendIdmapSources := c.shouldSendIdmapSources()
	if !sendMountSources && !sendIdmapSources {
		return nil
	}

	// Elements on these slices will be paired with mounts (see StartInitialization() and
	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
	mountFds := make([]int, len(c.config.Mounts))
	idmapFds := make([]int, len(c.config.Mounts))
	for i, m := range c.config.Mounts {
		// The -1 fd is ignored later.
		mountFds[i] = -1
		idmapFds[i] = -1

		if !m.IsBind() {
			continue
		}

		// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
		// to allocate a fd so that we know the number to pass in the environment variable. The fd
		// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
		// lifecycle of that fd is already taken care of.
		cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)

		if sendMountSources && !m.IsIDMapped() {
			mountFds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
		if sendIdmapSources && m.IsIDMapped() {
			idmapFds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
	}

	if sendMountSources {
		if err := cmdAddEnvFds(cmd, "_LIBCONTAINER_MOUNT_FDS", mountFds); err != nil {
			return err
		}
	}
	if sendIdmapSources {
		if err := cmdAddEnvFds(cmd, "_LIBCONTAINER_IDMAP_FDS", idmapFds); err != nil {
			return err
		}
	}

	return nil
}
```

Looking at that; perhaps we're trying to do too many things in this function; effectively there's two separate things happening;

- setting `_LIBCONTAINER_MOUNT_FDS` (if needed)
- setting `_LIBCONTAINER_IDMAP_FDS` (if needed)


And they both are mostly separate.

thaJeztah replied:
> And they both are mostly separate.

Assuming that the order in which they get added in `cmd.ExtraFiles` doesn't matter (but let me know if they do ⚠️ ), perhaps split them to separate methods; thinking along these lines

```go
func (c *Container) sendSources(cmd *exec.Cmd, messageSockPair filePair) error {
	if err := c.sendMountSources(cmd, messageSockPair); err != nil {
		return err
	}
	if err := c.sendIdmapSources(cmd, messageSockPair); err != nil {
		return err
	}
	return nil
}

func (c *Container) sendMountSources(cmd *exec.Cmd, messageSockPair filePair) error {
	if !c.shouldSendMountSources() {
		return nil
	}

	// Elements on these slices will be paired with mounts (see StartInitialization() and
	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
	fds := make([]int, len(c.config.Mounts))
	for i, m := range c.config.Mounts {
		// The -1 fd is ignored later.
		fds[i] = -1

		if !m.IsBind() {
			continue
		}

		// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
		// to allocate a fd so that we know the number to pass in the environment variable. The fd
		// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
		// lifecycle of that fd is already taken care of.
		cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)

		if !m.IsIDMapped() {
			fds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
	}

	return cmdAddEnvFds(cmd, "_LIBCONTAINER_MOUNT_FDS", fds)
}

func (c *Container) sendIdmapSources(cmd *exec.Cmd, messageSockPair filePair) error {
	if !c.shouldSendIdmapSources() {
		return nil
	}

	// Elements on these slices will be paired with mounts (see StartInitialization() and
	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
	fds := make([]int, len(c.config.Mounts))
	for i, m := range c.config.Mounts {
		// The -1 fd is ignored later.
		fds[i] = -1

		if !m.IsBind() {
			continue
		}

		// The fd passed here will not be used: nsexec.c will overwrite it with dup3(). We just need
		// to allocate a fd so that we know the number to pass in the environment variable. The fd
		// must not be closed before cmd.Start(), so we reuse messageSockPair.child because the
		// lifecycle of that fd is already taken care of.
		cmd.ExtraFiles = append(cmd.ExtraFiles, messageSockPair.child)
		if m.IsIDMapped() {
			fds[i] = stdioFdCount + len(cmd.ExtraFiles) - 1
		}
	}

	return cmdAddEnvFds(cmd, "_LIBCONTAINER_IDMAP_FDS", fds)
}
```

rata replied:
@thaJeztah I've written as different methods before and didn't like it much either, that is why I switched to this. Do you have a strong opinion? Let me know and I can change to match your taste, I didn't like much any of these, but decided for this version in the end

thaJeztah replied:
I'll let others add their view as well, but my thinking here was that the separate methods may make the logic less ambiguous.

Looking at `sendIdMapSources` I can see that 

- if there's no need to send them, we return early (`! c.shouldSendIdmapSources()`)
- we're only interested in `bind` mounts
- if that bind-mount is idmapped, we're adding it to `_LIBCONTAINER_IDMAP_FDS`

rata replied:
Yeap, to be honest, I was pretty much on the fence between these two options. I'll update it to have them be two different methods, then!

rata replied:
@thaJeztah changed it to be different cases and different functions. But instead of just complete c&p, I've used a common function (sendFdsSources) that takes the condition as a func param.

I think this abstraction is way simpler and acceptable. But if you prefer to have it all c&p, please let me know.

Closing this, though, as it is probably better to comment in the new code (this doesn't exist). Feel free to re-open if I'm missing something

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
+	idmapFds := make([]int, len(c.config.Mounts))
+	for i, m := range c.config.Mounts {
+		// The -1 fd is ignored later.
+		mountFds[i] = -1
```
Not critical, but my IDE is warning that  `mountFds` here is shadowing the `mountFds` type that was added in libcontainer/init_linux.go. Perhaps best to pick a different name for one of them.

rata replied:
I don't see this in my vim, but I think it might be solved as this var doesn't exist anymore. But please confirm :)

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -2222,7 +2281,7 @@ func (c *Container) bootstrapData(cloneFlags uintptr, nsMaps map[configs.Namespa
 		var mounts []byte
 		for _, m := range c.config.Mounts {
-			if m.IsBind() {
+			if m.IsBindOnly() {
```
```suggestion
			if m.IsBind() && !m.IsIDMapped() {
```

rata replied:
idem

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -2237,6 +2296,28 @@ func (c *Container) bootstrapData(cloneFlags uintptr, nsMaps map[configs.Namespa
+	if it == initStandard && c.shouldSendIdmapSources() {
+		var mounts []byte
+		for _, m := range c.config.Mounts {
+			if m.IsIDMap() {
```
for consideration (see other comments)

```suggestion
			if m.IsBind() && m.IsIDMapped() {
```

rata commented on libcontainer/configs/mount_linux.go:
```diff
@@ -29,8 +31,21 @@ type Mount struct {
+	return m.IsBind() && !m.IsIDMap()
+}
+
+func (m *Mount) IsIDMap() bool {
+	return m.IsBind() && (len(m.UIDMappings) > 0 || len(m.GIDMappings) > 0)
```
@thaJeztah The only way to create an idmap mount is creating a bind-mount. You can't just mount anything with ID translation, you can only do a bind-mount and add ID mappings to it.

So, do you think is ok if I only rename the function name?

brauner replied:
That's not entirely correct. :) In the new mount api there's no distinction between an fd gotten from `open_tree(OPEN_TREE_CLONE)` and one gotten from `fsmount()`. Both are `O_PATH` fds that refer to an anonymous mount or mount tree. So:
```
fd_mnt = open_tree("/mnt", OPEN_TREE_CLONE);
mount_setattr(fd_mnt, ...);
```
and

```
fd_fs = fsopen("tmpfs);
fsconfig(FSCONFIG_CMD_CREATE);
fd_mnt = fsmount(fd_fs);
mount_setattr(fd_mnt, ...);
```

both work.

rata replied:
@brauner ohh, interesting, thanks! We definitely don't support this in the code, though (we are using open_tree() and just that). 

Then changing this should be more correct, but we will have to check if it's a bind mount too (we can probably validate it earlier, as otherwise it is not supported).

@thaJeztah What do you think about doing that?

rata commented on libcontainer/container_linux.go:
```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
+	// prepareRootfs()). These slices MUST have the same size as c.config.Mounts.
+	mountFds := make([]int, len(c.config.Mounts))
+	idmapFds := make([]int, len(c.config.Mounts))
+	for i, m := range c.config.Mounts {
```
@thaJeztah yes, that is true. I expect the compiler, though, to easily avoid computing the boolean result of `c.shouldSendMountSources()` and `c.shouldSendIdmapSources()`.

Do you think the compiler isn't smart enough and we should store this in a variable?

rata replied:
As we talked in other comments, I've re-worked this so the cases are separate and this issue doesn't exist anymore. I think, at least this comment, is solved :)

rata commented on libcontainer/container_linux.go:
```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
+	// We need to send sources if there are non-idmap bind-mounts.
 	for _, m := range c.config.Mounts {
-		if m.IsBind() {
+		if m.IsBindOnly() {
```
@thaJeztah This is the reason I was saying I don't really like this option. If tomorrow there is another kind of mount that we support, we will need to change in tons of other places this to: `if m.IsBind() && !m.IsIDMapped() && !m.IsXXX()`

And not only here, of course.

What we really want to know here is if this is a plain/vanilla/simple bind mount. I don't see the point of duplicating these checks throughout the code. If we want to check if this is a simple bind mount, why not have a method that checks that?

thaJeztah replied:
We can't look into the future, but I'm always a bit wary of trying to abstract things too much. Especially with multiple boolean operators combined (some of which are only "tangential" related). We may try to hide complexity here, but at the same time add additional cognitive load, or potentially have to dig ourselves in further (`IsOtherTypeMountOnly()`, `IsOtherTypeMountWithIDMap()`)

rata replied:
@thaJeztah I agree too many abstractions is not good. I'll update to just open-code this and remove this method, then!

rata replied:
Done!

rata commented on libcontainer/container_linux.go:
```diff
@@ -518,57 +518,116 @@ func (c *Container) shouldSendMountSources() bool {
+
+	// We need to send sources if there are idmap bind-mounts.
+	for _, m := range c.config.Mounts {
+		if m.IsIDMap() {
```
Done

rata commented:
@thaJeztah @AkihiroSuda Fixed all review comments now, PTAL

AkihiroSuda commented on libcontainer/nsenter/idmap.h:
```diff
@@ -0,0 +1,78 @@
+		#define __NR_mount_setattr (442 + 1024)
+	#else
+		#define __NR_mount_setattr 442
+	#endif
```
nit: Go won't support Alpha and IA64

rata replied:
@AkihiroSuda I removed it. But now it is `if mips <mips> else #define __NR_mount_setattr 442`, which can be tricky to debug if go adds another arch that has a different syscall number.

Given your comment and that adding support for alpha or ia64 seems unlikely, This is probably safe. But let me know if you change your mind now that it looks like how I mentioned :)

AkihiroSuda commented on tests/integration/helpers.bash:
```diff
@@ -620,3 +621,36 @@ function requires_kernel() {
+			#	https://lore.kernel.org/lkml/20210510125147.tkgeurcindldiwxg@wittgenstein
+			#
+			# So, sadly we just need to skip this on centos.
+			skip "sadly, for no technical reason centos kernel 5.14 does not permit using ID map mounts"
```
```suggestion
			skip "sadly, centos kernel 5.14 does not permit using ID map mounts"
```
might be less controversial

rata replied:
Applied, thanks

AkihiroSuda commented:
https://github.com/opencontainers/runc/blob/main/docs/spec-conformance.md should be updated too (can be another PR)

rata commented:
@AkihiroSuda thanks! Opened the other PR and did the small fixes you suggested. But the push for that made it lose your LGTM. Can you please re-add your LGTM? :)

rata commented:
Friendly ping? @kolyshkin @thaJeztah

rata commented:
I solved the conflict that was created with the merge a few minutes ago using the github web interface, that is why it added a merge. I can squash it later, but it should be fine to review anyways :)

AkihiroSuda commented:
> [Merge branch 'main' into rata/idmap](https://github.com/opencontainers/runc/pull/3717/commits/1422e62fd4c5cc5a977c58d6f8a2f2928531f09e)

Still LGTM but please clean up the commit log

rata commented:
@AkihiroSuda thanks, fixed!

rata commented:
@kolyshkin @thaJeztah Now that runc 1.1.5 is out, can you PTAL?

Rebased to see if the centos-9 issue is fixed (it does not seem not related to this PR)

rata commented:
CI seems kind of broken, but unrelated to this PR. I guess the open PR from @kolyshkin is taking care of that.

kolyshkin commented on libcontainer/configs/validate/validator.go:
```diff
@@ -259,6 +260,39 @@ func cgroupsCheck(config *configs.Config) error {
+		if !sameMapping(config.GidMappings, m.GIDMappings) {
+			return fmt.Errorf("invalid mount %+v: not yet supported for the mount gidMappings to be different than user namespace gidMapping", m)
+		}
+		// For idmap mounts we require src and dst to be an absolute path.
+		if !filepath.IsAbs(m.Destination) {
+			return fmt.Errorf("invalid mount %+v: mount destination not absolute", m)
+		}
```
We have the same check in mounts() check, which was downgraded to a warning by commit b31a9340f92bf02d7fc1a8a9513885561e2fbc47. As this was done 2 years ago, maybe it's time to act on this TODO item and upgrade it to a warning, in the anticipation of 1.2.0?

rata replied:
@kolyshkin s/upgrade to a warning/upgrade to an error, I guess?

rata replied:
Upgraded to an error

kolyshkin commented:
Is it possible to rework the "Support idmap mounts on volumes" commit into two?
1. Generalize sending mount sources (i.e. refactor the code, do something similar to what you did in "nsexec: Add generic receive_sources()")
2. Add new functionality.

It is more logical and easier to review that way

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -259,6 +260,39 @@ func cgroupsCheck(config *configs.Config) error {
+			continue
+		}
+
+		if !m.IsBind() {
+			return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is supported only for mounts with the option 'bind'", m)
+		}
+		if config.RootlessEUID {
+			return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported when runc is being launched with EUID != 0, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
+		}
+		if config.UidMappings == nil || config.GidMappings == nil {
+			return fmt.Errorf("invalid mount %+v: not yet supported to use GidMappings/UidMappings in a mount without also using a user namespace", m)
+		}
+		if !sameMapping(config.UidMappings, m.UIDMappings) {
+			return fmt.Errorf("invalid mount %+v: not yet supported for the mount uidMappings to be different than user namespace uidMapping", m)
+		}
+		if !sameMapping(config.GidMappings, m.GIDMappings) {
+			return fmt.Errorf("invalid mount %+v: not yet supported for the mount gidMappings to be different than user namespace gidMapping", m)
+		}
```
I thought the same. But other errors are already printing the whole struct and my thinking was: most users don't use runc directly, they use it via containerd or kubernetes, and in those cases it is not easy to see what the config.json has for a user. So keeping the complete error message will help us to know what is wrong and debug users bug reports (like using X flag and some other option that is not allowed).

So I thought it was better to keep the whole struct. It should really never happen (so not possible this will be spamming logs nor anything, these are errors that stop the execution) and whenever they do happen, better to have all the info.

What do you think?

rata replied:
I take that reaction as an ok, so I kept this

rata commented:
Rebased on top of latest main, to see if the CI is now fixed

rata commented:
CI is fixed! :-D

rata commented:
@kolyshkin I've upgraded that to an error, PTAL

rata commented:
I've also added unit tests for the validator. @kolyshkin @thaJeztah PTAL

AkihiroSuda commented:
> Expected — Waiting for status to be reported

Sorry, could you try rebasing again

rata commented:
@AkihiroSuda rebased, hipe the CI doesn't get stuck now

rata commented:
CI run fine now, all is green :)

rata commented:
Rebased to fix conflicts

@kolyshkin Can you please take a look?

rata commented:
Centos failure is unrelated, repushing

rata commented:
All green, PTAL

kolyshkin commented on libcontainer/configs/validate/validator.go:
```diff
@@ -253,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+	}
+
+	if !m.IsBind() {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is supported only for mounts with the option 'bind'", m)
```
Overall, I don't like printing out the whole `struct Mount` in an error message.

Can a mount be uniquely identified by its destination?

Same for similar errors below.

kolyshkin replied:
Makes sense. Sorry for asking the same q over and over :)

kolyshkin commented on libcontainer/specconv/spec_linux.go:
```diff
@@ -497,6 +497,18 @@ func CreateLibcontainerConfig(opts *CreateOpts) (*configs.Config, error) {
 	return config, nil
 }
 
+func toConfigIDMap(specMaps []specs.LinuxIDMapping) []configs.IDMap {
```
I think it makes sense to also switch `setupUserNamespace` to using this function. This may be done by a separate commit just before this one.

rata replied:
Good call, switched that. Thanks

kolyshkin commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -740,40 +740,49 @@ void send_fd(int sockfd, int fd)
+ * not -1, it receives an fd via SCM_RIGHTS and dup3 it to the fd requested in
+ * the element of the env var.
+ */
+void receive_fd_sources(int sockfd, char *env_var)
```
nit: maybe `const char *env_var`?

rata replied:
Fixed, thanks

kolyshkin commented on libcontainer/nsenter/idmap.h:
```diff
@@ -0,0 +1,74 @@
+#include <sys/mount.h>
+// Centos-7 doesn't have this file nor the __has_include() directive, so let's
+// just leave this commented out and we can uncomment when it hits EOL (2024-06-30).
+//#include <linux/mount.h>
```
So, I think this should be handled slightly differently. 

First, no need to `#include <linux/mount.h>`, as everything seems to be coming from other include files.

Second, maybe don't define syscall numbers if they don't exist on a platform used for compilation, instead doing something

```c
#if defined(MOUNT_ATTR_IDMAP) && defined(__NR_mount_setattr) && defined(__NR_open_tree)
#define HAVE_IDMAP_MOUNT
#else
#warning "not compiling in support for idmapped mounts"
#endif
```

and using `#ifdef HAVE_IDMAP_MOUNT` later in the code.

kolyshkin replied:
I mean, have something like in a commit that adds the functionality
```c
void send_idmapsources(int sockfd, pid_t pid, char *idmap_src, int idmap_src_len)
#ifndef HAVE_IDMAP_MOUNT
{
    bail("no support for idmapped mounts");
}
#else
{
....
}
#endif /* HAVE_IDMAP_MOUNT */
```

rata replied:
@kolyshkin regarding the first point, no, not everything is coming from other includes. Look at the mount setattr manpage: https://man7.org/linux/man-pages/man2/mount_setattr.2.html:

```
       #include <linux/mount.h> /* Definition of MOUNT_ATTR_* constants */
```

What helps us to drop that include and avoid hitting the issues the comment says is that we _do_ these defines here. So the define for `MOUNT_ATTR_IDMAP` comes from this file and we can drop the include to just fix the issues in old centos.

And centos-9 was also broken but in a more weird way (probably due to the patch to disable idmap mounts there, that is now being reverted) and I needed to add another ifdef here: https://github.com/opencontainers/runc/pull/3717/files#diff-421c257af6d93ef0bbd742757244adec93587030041c3306af559a1359e35532R27-R28 (centos-9 for some reason had __NR_mount_setattr defined but not the struct to pass the parameters, which is just so broken).

All of this to explain that supporting so many old OSes makes some things not ideal and it is not obvious if something like that will work. I'll try to see if there is a reasonable way to do something like you suggest. But this takes time, as the CI is slow and this runs in vagrant that is WAY more slower to run and painful to debug.

rata replied:
@kolyshkin didn't manage to do anything good in this front.

I pushed what I was trying to this branch: https://github.com/kinvolk/runc/tree/rata/idmap-simpler-headers

And explained the problem there, but will put it here just in case:

 We can't avoid the defines of __NR_open_tree and __NR_mount_setattr.
 The thing is:
  * If we include linux/mount.h, then compilation on centos-7 fails (doesn't have that file). We can't
  use the __has_include() directive there as it seems the gcc version used
  doesn't support it.
  * So we need to not include linux/mount.h. But that means that we need to
  define the "struct mount_attr" and the constants: MOUNT_ATTR_IDMAP,
  OPEN_TREE_CLONE, OPEN_TREE_CLOEXEC. But sadly, even doing that is not
  enough, as then __NR_open_tree nor __NR_mount_setattr isn't defined either.

  So, basically, it seems we need to keep the defines here as long as we
  support centos-7.

Sounds good then to keep this as it is?

rata replied:
Maybe there is something else that can be done for the first item (make centos-7 don't complain with that include), but I haven't found it.

Also, I was playing with this again today and I found another barrier to remove the defines of the syscalls. I don't think we can remove them anytime soon :(.

Ubuntu 20.04 ships with linux-libc-dev version 5.4.0-147.164 (or 5.4.whatever). This version was built for the 5.4 kernel headers, that was originally shipped with Ubuntu 20.04, and include open_tree but does include mount_sysattr. However, Ubuntu upgraded the kernel to 5.15 in this release (the github action uses a 5.15 kernel too, see this for the focal kernel package https://packages.ubuntu.com/search?keywords=linux-image-5.15.0-71-generic). It probably makes sense to keep the 5.4 headers for Ubuntu, so apps are unchanged, and just ship an updated kernel.

Therefore, we want to run on Ubuntu 20.04, but ubuntu ships with outdated headers. So we want to use our defines and make this feature just work on kernels that have the features we need.

I think this should solve the general concerns about the defines, as it really seems there is no way around it unless we drop support for lot of popular distros that are to be maintained several more years. I strongly think we should keep the defines now that we know this.

@kolyshkin what do you think?

rata replied:
@kolyshkin Also, I added this: https://github.com/opencontainers/runc/pull/3717/files#diff-6383238247e090d88ade6343c0ef59dd09b3c10634bf0584e78445b843c55ab0R913-R914

I was already doing that for `mount_setattr(2)` but not for `open_tree(2)`. I think handling it this way has the same advantages than the version you proposed that just bailed.

So, hopefully this can be marked as solved @kolyshkin ? Please resolve if you agree :)

kolyshkin replied:
Sounds good to me.

kolyshkin commented on libcontainer/init_linux.go:
```diff
@@ -46,6 +46,15 @@ type network struct {
 }
 
+type mountFds struct {
+	// Fds to use as source when mounting
```
nit: the comment should start with `sourceFds is/are ...`

rata replied:
Fixed, thanks.

We probably need a linter for these things

kolyshkin commented on libcontainer/init_linux.go:
```diff
@@ -75,7 +84,7 @@ type initer interface {
 }
 
-func newContainerInit(t initType, pipe *os.File, consoleSocket *os.File, fifoFd, logFd int, mountFds []int) (initer, error) {
+func newContainerInit(t initType, pipe *os.File, consoleSocket *os.File, fifoFd, logFd int, mountFds mountFds) (initer, error) {
```
Should we pass the structure by the reference instead?

kolyshkin commented on libcontainer/container_linux.go:
```diff
@@ -531,6 +531,38 @@ func (c *Container) shouldSendMountSources() bool {
+	//
+	// Let's just rule out rootless, we don't have those permission in the
+	// rootless case.
+	if c.config.RootlessEUID {
```
nit: can we put this check first, as this is the easiest/fastest one?

rata replied:
Sure, done.

kolyshkin commented on libcontainer/rootfs_linux.go:
```diff
@@ -463,8 +479,35 @@ func mountToRootfs(m *configs.Mount, c *mountConfig) error {
+
+		if m.IsBind() && m.IsIDMapped() {
+			if idmapFd == nil {
+				return fmt.Errorf("error creating mount %+v: idmapFd is nil, should point to a valid fd", m)
```
Same here -- I don't like printing out the whole nine yards of mount structure in an error message. Perhaps it can be identified by its destination?

rata replied:
Ok. Let's centralize the discussion on the other thread and then we apply what we agreed here too

kolyshkin commented on libcontainer/rootfs_linux.go:
```diff
@@ -463,8 +479,35 @@ func mountToRootfs(m *configs.Mount, c *mountConfig) error {
+				return fmt.Errorf("error creating mount %+v: idmapFd is nil, should point to a valid fd", m)
+			}
+			if err = unix.MoveMount(*idmapFd, "", -int(unix.EBADF), dest, unix.MOVE_MOUNT_F_EMPTY_PATH); err != nil {
+				return fmt.Errorf("error on unix.MoveMount %+v: %w", m, err)
```
ditto. Also, maybe use `&os.PathError`

rata replied:
Ok. Let's centralize the discussion on the other thread and then we apply what we agreed here too

kolyshkin commented on libcontainer/rootfs_linux.go:
```diff
@@ -77,6 +82,16 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds mountFds) (
 		}
 
+		// We validated before we can access idmapFds[i].
+		mountConfig.idmapFd = nil
```
No need to explicitly set it to `nil`, as this is the default value.

rata replied:
@kolyshkin nope, we need to. The mount config struct is created outside the loop, so if we don't re-set it to nil on each operation, we can use an old fd.

This actually caused a bug that you fixed by setting it the mountConfig.fd to nil in the else case: https://github.com/opencontainers/runc/commit/d370e3c04660201e72ba6968342ce964c31a2d7f

The fix I proposed that you didn't like was moving the creation of the mountConfig struct inside the loop, so we don't need to care about re-setting it. But you didn't like that, so we need to assign nil when the mount doesn't have an fd.

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -253,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+	if config.RootlessEUID {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported when runc is being launched with EUID != 0, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
+	}
+	if config.UidMappings == nil || config.GidMappings == nil {
```
Great point, fixed

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -35,6 +35,7 @@ type mountConfig struct {
 	rootlessCgroups bool
 	cgroupns        bool
 	fd              *int
+	idmapFd         *int
```
I'm unsure if we want to have fd and idmapFd or just have one fd field here and when the mount is idmapped it is the idmap fd, when it is a mountFd then the fd field is that and when none is needed is nil.

@kolyshkin what do you think?

kolyshkin replied:
First of all, there should not be any fd here, since `mountConfig` is not about a particular mount. This was fixed by commit a60933bb245657e3e1f3 (and you need to change your code).

I think that we can and should reuse `mountEntry.srcFD` field.

rata replied:
This was already done by Francis. It seems reusing srcFD is not straight forward as that is a string and we need an int for the idmap case.

Even if we use a string, we need to know when it is a srcFD, so we add /proc/self/fd/ and in the idmap case we don't need to do that.

There might be a way to do it, but it seems error-prone. If you disagree @kolyshkin , please re-open :)

rata commented on tests/integration/helpers.bash:
```diff
@@ -616,3 +617,36 @@ function requires_kernel() {
+	local fs
+	fs=$1
+
+	# We need to true it to avoid CI failure as this binary may return with
```
No, it was meant to say "|| true" the command. 

Changed it to:
```
        # We need to "|| true" it to avoid CI failure as this binary may return with
        # something different than 0.
```

rata commented:
@kolyshkin fixed all, except for the nsexec parts. PTAL I'll do the nsexec parts now, but please let me know if the rest is ready to go.

rata commented:
Rebased on top of latest main, but should still be all green

rata commented:
@kolyshkin All concerns should be solved (or answered) now. Also, rebased on latest main and all test should be green.

As a summary: the nsexec changes you mentioned are not really possible (let's continue on that thread if you want more info).

rata commented:
Pushing again to restart the CI, it was stalled without starting to run the tests on only one of the checks. It should be all green, at least if the CI doesn't hang :)

rata commented:
Rebased on latest main. @kolyshkin FYI, I'll be afk next week. If we can merge this today or tomorrow, it would be great. Let me know what you think :)

rata commented:
@kolyshkin ping?

AkihiroSuda commented:
Needs rebase

kolyshkin commented on libcontainer/container_linux.go:
```diff
@@ -531,6 +531,38 @@ func (c *Container) shouldSendMountSources() bool {
+// the mount_setattr already done in the host user namespace.
+func (c *Container) shouldSendIdmapSources() bool {
+	// nsexec.c mount_setattr() requires CAP_SYS_ADMIN in:
+	// * the user namespace the filesystem was mounted in
+	// * the user namespace we're trying to idmap the mount to
```
nit: missing `;` at EOL

rata replied:
Fixed!

kolyshkin commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -861,6 +875,83 @@ void try_unshare(int flags, const char *msg)
+	/* Open the userns fd only once.
+	 * Currently we only support idmap mounts that use the same mapping than
+	 * the userns. This is validated in libcontainer/configs/validate/validator.go,
+	 * so if we reached here, we know the mapping for the idmap is the same
+	 * than the userns. This is why we just open the userns_fd once from the
```
`s/is the same than/is the same as/`?

kolyshkin commented on libcontainer/standard_init_linux.go:
```diff
@@ -86,15 +86,15 @@ func (l *linuxStandardInit) Init() error {
+	// We don't need the mount nor idmap fds after prepareRootfs() nor if it fails.
 	err := prepareRootfs(l.pipe, l.config, l.mountFds)
-	for _, m := range l.mountFds.sourceFds {
+	for _, m := range append(l.mountFds.sourceFds, l.mountFds.idmapFds...) {
```
This creates another slice. Not a big deal I guess.

rata replied:
Closing this

kolyshkin commented on tests/integration/helpers.bash:
```diff
@@ -616,3 +617,36 @@ function requires_kernel() {
+		skip "$fs underlying file system does not support ID map mounts"
+		;;
+	*operation\ not\ permitted)
+		if uname -r | grep -q el9; then
+			# centos kernel 5.14.0-200 does not permit using ID map mounts due to a
+			# specific patch added to their sources:
+			# 	https://gitlab.com/redhat/centos-stream/src/kernel/centos-stream-9/-/merge_requests/131
+			#
+			# There doesn't seem to be any technical reason behind
+			# it, none was provided in numerous examples, like:
+			# 	https://lore.kernel.org/lkml/20210213130042.828076-1-christian.brauner@ubuntu.com/T/#m3a9df31aa183e8797c70bc193040adfd601399ad
+			#	https://lore.kernel.org/lkml/20210213130042.828076-1-christian.brauner@ubuntu.com/T/#m59cdad9630d5a279aeecd0c1f117115144bc15eb
+			#	https://lore.kernel.org/lkml/m1r1ifzf8x.fsf@fess.ebiederm.org
+			#	https://lore.kernel.org/lkml/20210510125147.tkgeurcindldiwxg@wittgenstein
+			#
+			# So, sadly we just need to skip this on centos.
+			skip "sadly, centos kernel 5.14 does not permit using ID map mounts"
+		fi
```
Was the mentioned patch reverted? If yes, maybe we can remove this kludge.

kolyshkin replied:
Maybe add a TODO item to re-enable the test?

rata replied:
No need to remove anything, we only skip it if when we try an idmap mount in the fs, it returns EPERM and we are running on centos. This should just work as is if we run with a new centos that doesn't return EPERM :)

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,227 @@
+load helpers
+
+function setup() {
+	# idmap mounts requires root and kernel >= 5.12
```
Comments should explain not what we are doing, but why we're doing that. In here, it is obvious from the below two lines that we require this and that. So I guess this comment can be safely removed.

kolyshkin commented:
LGTM overall, just a few nits, and it needs to be rebased.

thaJeztah commented on libcontainer/configs/validate/validator_test.go:
```diff
@@ -387,3 +386,195 @@ func TestValidateMounts(t *testing.T) {
+		},
+	}
+
+	for _, tc := range testCases {
+		config := tc.config
+		config.Rootfs = "/var"
+
+		err := mounts(config)
+		if tc.isErr && err == nil {
+			t.Errorf("mount test: %v, expected error, got nil", tc.name)
+		}
+		if !tc.isErr && err != nil {
+			t.Errorf("mount config: %v, expected nil, got error %q", tc.name, err)
+		}
+	}
+}
```
Might as well make this actual sub-tests; in that case, it's also not needed to include the test-name in the `t.Errorf(..)`

```suggestion
	for _, tc := range testCases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			config := tc.config
			config.Rootfs = "/var"

			err := mounts(config)
			if tc.isErr && err == nil {
				t.Error("expected error, got nil")
			}
			if !tc.isErr && err != nil {
				t.Error(err)
			}
		})
	}
}
```

thaJeztah commented:
Ah, yes, we've had that happen sometimes. I gave it another kick; let's hope it was just a short glitch 🤞

thaJeztah commented:
> I will provide cover while rata is in sick leave.

Hope he's doing ok, and that it's nothing serious. In either case; if you happen to connect with him, wish him "get well soon 💛"

kolyshkin commented on libcontainer/configs/mount_linux.go:
```diff
@@ -29,8 +29,24 @@ type Mount struct {
+	// Note that, the underlying filesystem should support this feature to be
+	// used.
+	// Every mount point could have its own mapping.
+	GIDMappings []IDMap `json:"gidMappings"`
```
Perhaps we should add `omit_empty`? Same for UIDMappings above

rata replied:
This was solved already :)

kolyshkin commented on libcontainer/rootfs_linux.go:
```diff
@@ -40,7 +40,8 @@ type mountConfig struct {
 	*configs.Mount
-	srcFD string
+	srcFD   string
+	idmapFd *int
```
Hmm. Can we reuse srcFD here?

kolyshkin replied:
I guess not

kolyshkin commented on libcontainer/rootfs_linux.go:
```diff
@@ -470,8 +484,35 @@ func mountToRootfs(c *mountConfig, m mountEntry) error {
+			if m.idmapFd == nil {
+				return fmt.Errorf("error creating mount %+v: idmapFd is nil, should point to a valid fd", m)
+			}
+			if err = unix.MoveMount(*m.idmapFd, "", -int(unix.EBADF), dest, unix.MOVE_MOUNT_F_EMPTY_PATH); err != nil {
```
This is a weird use of an errno value. From what I understand, `dest` is absolute path and so `destFd` argument is ignored. You can probably just set it to -1.

rata replied:
This was solved already :)

kolyshkin commented on .gitignore:
```diff
@@ -4,6 +4,7 @@ vendor/pkg
 contrib/cmd/recvtty/recvtty
 contrib/cmd/sd-helper/sd-helper
 contrib/cmd/seccompagent/seccompagent
+contrib/cmd/fs_idmap/fs_idmap
```
nit: can you please use `-` not `_`, just for the uniformity sake (see `sd-helper` above)?

rata replied:
This was solved already :)

kolyshkin commented on contrib/cmd/fs_idmap/fs_idmap.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	src := os.Args[1]
+	treeFD, err := unix.OpenTree(-int(unix.EBADF), src, uint(unix.OPEN_TREE_CLONE|unix.OPEN_TREE_CLOEXEC|unix.AT_EMPTY_PATH|unix.AT_RECURSIVE))
```
Same weird use of errno code. You can use `-1` here I think.

rata replied:
This suggestion slipped through the cracks, fixed now!

kolyshkin commented on tests/integration/helpers.bash:
```diff
@@ -616,3 +617,42 @@ function requires_kernel() {
+
+	# We need to "|| true" it to avoid CI failure as this binary may return with
+	# something different than 0.
+	stderr=$($FS_IDMAP "$fs" 2>&1 >/dev/null || true)
```
In shell, `2>&1 >/dev/null` is not working as expected. You have to use `>/dev/null 2>&1`.

kolyshkin commented on libcontainer/rootfs_linux.go:
```diff
@@ -40,7 +40,8 @@ type mountConfig struct {
 	*configs.Mount
-	srcFD string
+	srcFD   string
+	idmapFd *int
```
not very important, but we use `srcFD` so I guess the proper name is `idmapFD`.

rata replied:
This was solved already :)

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,226 @@
+
+@test "idmap mount with propagation flag" {
+	update_config '   .process.args = ["sh", "-c", "findmnt -o PROPAGATION /tmp/mount-2"]
+			| .mounts += [
+					{
+						"source": "source-2/",
+						"destination": "/tmp/mount-2",
+						"options": ["bind", "shared"],
+						"uidMappings": [ {
+						                  "containerID": 0,
+						                  "hostID": 100000,
+						                  "size": 65536
+						                }
+						],
+						"gidMappings": [        {
+						                  "containerID": 0,
+						                  "hostID": 100000,
+						                  "size": 65536
+						                }
+						]
+					}
+				] '
```
It would be much better if, instead of adding a new mount, you can modify the one added in `setup`, do add `shared` to options. This way, the test will be way more readable, as it's clear what we are changing and testing here.

rata replied:
Sure. I didn't do it like that because I don't like much the jq syntax nor having a phd in jq. But adjusted it now :)

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,226 @@
+					{
+						"source": "source-2/",
+						"destination": "/tmp/mount-2",
+						"options": ["bind"]
```
Same here -- I guess you just need to remove uid/gid mappings from the mount.

rata replied:
Hmm, no, here I want to try a config that has an idmap mount AND a regular bind mount (to make sure the fd passing mechanism doesn't screw it up when both type of sources are present)

I want the two entries, and the only similarity is that their type is "bind". I think a new completely different entry is more readable

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,226 @@
+						"source": "source-2/",
+						"destination": "/tmp/mount-2",
+						"options": ["bind"],
+						"uidMappings": [ {
```
same here

rata replied:
Fixed

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,226 @@
+						"source": "source-2/",
+						"destination": "/tmp/mount-2",
+						"options": ["bind"],
+						"gidMappings": [ {
```
same here

rata replied:
Fixed

kolyshkin commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,226 @@
+	update_config '   .mounts += [
+					{
+						"source": "tmpfs",
+						"destination": "/tmp/mount-2",
```
same here

kolyshkin commented:
Left a few comments, mostly nits. Hope we can merge this soon

rata commented on libcontainer/configs/mount_linux.go:
```diff
@@ -1,6 +1,8 @@
-import "golang.org/x/sys/unix"
+import (
+	"golang.org/x/sys/unix"
+)
```
This was solved already :)

rata commented on libcontainer/rootfs_linux.go:
```diff
@@ -88,6 +93,15 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds mountFds) (
+		// We validated before we can access idmapFds[i].
+		if mountFds.idmapFds != nil && mountFds.idmapFds[i] != -1 {
+			entry.idmapFd = &mountFds.idmapFds[i]
+		}
```
This was solved already :)

rata commented:
@kolyshkin addressed all the remaining comments, PTAL
@AkihiroSuda PTAL

All tests are green except for `ci / cross-i386` that seems to explode when adding the criu ppa (that seems to still exist, not sure what might have broken that).

It took longer than I thought, but I'm back now and hopefully healing properly! :)

rata commented:
All tests should be green now. Just re-pushing without changes did the trick, it seems the failing job was something at github's end (it made sense, adding an apt repo should work).

centos-8 only still running now, but this was already passing before and there are no changes, it should just work if cirrus is not having issues.

Please take a look and, even better, merge :)

AkihiroSuda commented:
@kolyshkin Can we merge this?

rata commented:
@kolyshkin friendly ping?

Rebased on latest main, to fix conflicts with the ns-exec.c reworks from the PR merged yesterday: https://github.com/opencontainers/runc/pull/3599.

All tests should be green too :)

lifubang commented on tests/integration/idmap.bats:
```diff
@@ -0,0 +1,153 @@
+}
+
+@test "two idmap mounts with two bind mounts" {
+	update_config '   .process.args = ["sh", "-c", "stat -c =%u=%g= /tmp/mount-1/foo.txt /tmp/mount-2/foo.txt"]
```
How about add some test cases about checking the write permission in a container? For example, touch an existing file, create a new file in a container and check the owner in the host.

rata replied:
@lifubang Thanks! I've added a test for that, see the new "write to an idmap mount" test for more details.

I was on the fence if this was worth it or not, as it smells a little bit like testing the kernel feature rather than the runc feature, but the test is so simple and fast that I think there is no harm having it either.

rata replied:
@lifubang oh, I just realize you said "and check the owner in the host". IMHO, that seems more testing the kernel feature than the runc feature.

I added a test that we write a file and check its ownership from inside the container. As this is an idmap mount, the only way it is shown correctly is because the UID/GID are the right on the host too. IMHO, that is enough to check what we are doing on runc.

The kernel side already has lot of tests for idmap mounts in xfstests (there are several suites here: https://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git/tree/src/vfs, and there are some more tests in bash on that repo too).

But please feel free to re-open this if you think your concern is not addressed :)

rata commented:
@lifubang @kolyshkin PTAL :)

lifubang commented on libcontainer/configs/validate/validator.go:
```diff
@@ -262,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+	if len(config.UidMappings) == 0 || len(config.GidMappings) == 0 {
+		return fmt.Errorf("invalid mount %+v: not yet supported to use GidMappings/UidMappings in a mount without also using a user namespace", m)
+	}
+	if !sameMapping(config.UidMappings, m.UIDMappings) {
```
I read the discussions in `https://github.com/opencontainers/runtime-spec/pull/1143`.
I think this is too strict here. It will make the upstream projects can't specific the mount idmap params. I think maybe we should remove this check, let it check by kenel or upstream projects?
@thaJeztah @AkihiroSuda WDYT

rata replied:
I'm not sure what part of the discussion is relevant for this.

This is strict, yes, but this shortcut simplifies the initial implementation significantly as I mention in the PR description:
> This limitation is to
simplify the initial implementation, as all our users so far only need
this, and we can avoid sending over netlink the mappings, creating a
userns with this custom mapping, etc. Future PRs will remove this
limitation.

As I said there, this is all that all the upstream projects need so far (and we don't have plans to change that). This PR is already quite big as it is, I think it makes lot of sense to have this limitation in the initial implementation and lift it in follow-up PRs.

lifubang replied:
As you say, this PR is already quite big, I think it's quite big enough now. So I think to support idmap mount options may not make it too more bigger to review. But if you say this is all that all the upstream projects need so far, I have no other opinion.
LGTM except https://github.com/opencontainers/runc/pull/3717#discussion_r1259104260

rata replied:
Cool, I'll mark this as solved, then! :)

When I played with that, what wasn't trivial is changing the netlink protocol msg and then in C forking, creating a userns, killing that process after you used the mapping, etc. So I'm happy we can tackle that in a follow-up PR :)

lifubang commented on libcontainer/configs/validate/validator.go:
```diff
@@ -262,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+	if !sameMapping(config.UidMappings, m.UIDMappings) {
+		return fmt.Errorf("invalid mount %+v: not yet supported for the mount uidMappings to be different than user namespace uidMapping", m)
+	}
+	if !sameMapping(config.GidMappings, m.GIDMappings) {
```
The same here.

rata replied:
Let's unify the discussion in the other thread, we can change them in all the places later if we need to :)

lifubang commented on libcontainer/configs/validate/validator.go:
```diff
@@ -262,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
 func mounts(config *configs.Config) error {
 	for _, m := range config.Mounts {
+		// We upgraded this to an error in runc 1.2. We might need to
+		// revert this change if some users haven't still moved to use
```
As you say, we might need to revert this change. So let's double check.

rata replied:
Let's unify the discussion in the other thread.

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -29,21 +28,13 @@ func Validate(config *configs.Config) error {
 		sysctl,
 		intelrdtCheck,
 		rootlessEUIDCheck,
+		mounts,
```
The CHANGELOG file is usually created by the one doing the release. I've added a changelog entry comment to this PR description.

Please feel free to re-open if I'm missing something :)

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -29,21 +28,13 @@ func Validate(config *configs.Config) error {
 	}
-	// Relaxed validation rules for backward compatibility
-	warns := []check{
-		mounts, // TODO (runc v1.x.x): make this an error instead of a warning
```
See https://github.com/opencontainers/runc/issues/3020.

It should be safe, and it was requested by @kolyshkin, see the commit message: https://github.com/opencontainers/runc/pull/3717/commits/881e92a3fd5eeef80bac57efbb1c822deed4f895

rata commented:
@lifubang thanks for the review, PTAL :)

lifubang commented on libcontainer/standard_init_linux.go:
```diff
@@ -86,15 +86,15 @@ func (l *linuxStandardInit) Init() error {
 
 		if err := unix.Close(m); err != nil {
-			return fmt.Errorf("Unable to close mountFds fds: %w", err)
+			return fmt.Errorf("unable to close idmapFds and sourceFds: %w", err)
```
I think maybe we should keep the description of `mountFds fds`, otherwise it will be confused for someone who doesn't use idmap mounts.
Or change to close `sourceFds` and `idmapFds` seperatelly?

rata replied:
@lifubang  Right, makes sense! I've updated it to keep the current description. Thanks!

rata commented:
@lifubang fixed that now. Does it LGTY? :)

If you LGTM, @AkihiroSuda already LGTM, and all comments from any past review addressed, is this ready to merge? :scream: :tada: :crossed_fingers:

lifubang reviewed (approved):
LGTM, I think this can be merged, @AkihiroSuda . @cyphar @kolyshkin  Do you have some other opinions, because this PR is quite big.

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -596,6 +638,11 @@ func (c *Container) newInitProcess(p *Process, cmd *exec.Cmd, messageSockPair, l
 		return nil, err
 	}
 
+	// Send idmap sources.
```
Micro-nit (but in case there's other changes to make); I'd remove this comment, as it's not adding much value (the function is named `c.sendIdmapSources`, so it's describing the code, not the intent)

rata replied:
Fixed, thanks!

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -2259,6 +2306,29 @@ func (c *Container) bootstrapData(cloneFlags uintptr, nsMaps map[configs.Namespa
+		for _, m := range c.config.Mounts {
+			if m.IsBind() && m.IsIDMapped() {
+				// While other parts of the code check this too (like
+				// libcontainer/specconv/spec_linux.go) we do it here also beucause some libcontainer
```
Typo; `s/beucause/because/g`

thaJeztah commented on libcontainer/rootfs_linux.go:
```diff
@@ -73,6 +74,10 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds mountFds) (
 	}
 
+	if mountFds.idmapFds != nil && len(mountFds.idmapFds) != len(config.Mounts) {
+		return fmt.Errorf("malformed idmapFds slice. Expected size: %v, got: %v. Slice: %v", len(config.Mounts), len(mountFds.idmapFds), mountFds.idmapFds)
```
I see the existing error also does so, but I'm on the fence if printing `Slice: %v` is very useful, because the length of the size is the error, and the actual values in it not so much (?). The error message will just print a slice of int's (not sure if that helps debugging issues?).

As this is net-new code, perhaps also fix the "errors should not be capitalised of punctuation" to keep linters more happy; perhaps something like;

```go
return fmt.Errorf("malformed idmapFds slice: expected size: %v, got: %v", len(config.Mounts), len(mountFds.idmapFds))
```

thaJeztah commented on libcontainer/rootfs_linux.go:
```diff
@@ -40,7 +40,8 @@ type mountConfig struct {
 	*configs.Mount
-	srcFD string
+	srcFD   string
+	idmapFD int
```
I'm somewhat curious why we use a string for `srcFD` and an `int` for `idmapFD`. Wondering if we should unify the approach (probably can be done in a follow-up)

rata replied:
The srcFD is a string like "proc/self/fd/<int>" and we need it to be like that today. While the idmapFd is just an int, as we will just use it with moveMount later.

There might be some way to unify that, I'm not sure. Wanna create an issue, maybe? :)

thaJeztah replied:
> The srcFD is a string like "proc/self/fd/" and we need it to be like that today. While the idmapFd is just an int, as we will just use it with moveMount later.

Yup, noticed that 😄 

>  Wanna create an issue, maybe? :)

Sure! Done here; https://github.com/opencontainers/runc/issues/3935

rata replied:
Thanks!

thaJeztah commented on libcontainer/rootfs_linux.go:
```diff
@@ -81,13 +86,22 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds mountFds) (
+		}
+
+		if entry.idmapFD != -1 && entry.srcFD != "" {
+			return fmt.Errorf("malformed mountFds and idmapFds slice. Entry: %v has fds in both slices. One must be -1", i)
```
I guess the "One must be -1" is not fully correct, because `srcFD` is expected to be either `/proc/self/fd/<FD>` or empty 😅

rata replied:
haha, right! Great catch, I removed the last sentence and switched to use a comma before "Entry". Thanks!

thaJeztah commented on libcontainer/rootfs_linux.go:
```diff
@@ -466,8 +480,35 @@ func mountToRootfs(c *mountConfig, m mountEntry) error {
+			if m.idmapFD == -1 {
+				return fmt.Errorf("error creating mount %+v: idmapFD is invalid, should point to a valid fd", m)
+			}
+			if err = unix.MoveMount(m.idmapFD, "", -1, dest, unix.MOVE_MOUNT_F_EMPTY_PATH); err != nil {
```
Micro-nit; might want to make it explicit that `err` is local here;

```suggestion
			if err := unix.MoveMount(m.idmapFD, "", -1, dest, unix.MOVE_MOUNT_F_EMPTY_PATH); err != nil {
```

rata replied:
Fixed, thanks!

thaJeztah reviewed (commented):
Thanks! Overall this looks great. I left some comments/questions. I don't think there's huge blockers, and there's some _very_ tiny nits I left (but commented, in case things would be updated).

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -253,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is supported only for mounts with the option 'bind'", m)
+	}
+	if config.RootlessEUID {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported when runc is being launched with EUID != 0, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
```
We talked about this same thing with Kir here: https://github.com/opencontainers/runc/pull/3717#discussion_r1154707177

As I mentioned there, I understand this _can_ be too long, but it is also _super_ rare to hit. And whenever we hit it, is not that we can ask the user to see the config.json easily: k8s deletes it and retries super quick, it is not easy to find on which folder it is, etc.

And this is only hit in the case of users willing opting in for the feature.

I think having all the info about the mount, to understand what exactly we are dealing with, and maybe the info there can help us to see where in the code of other projects this config.json mount is being generated.


Did I convince you too? :)

thaJeztah replied:
> As I mentioned there, I understand this can be too long, but it is also super rare to hit. And whenever we hit it, is not that we can ask the user to see the config.json easily: k8s deletes it and retries super quick, it is not easy to find on which folder it is, etc.

Yup, not a blocker for me, and now that that part of the error is in a single location (where we wrap the error), it will be easier to update the formatting if we would decide to 👍

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -253,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported when runc is being launched with EUID != 0, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
+	}
+	if len(config.UidMappings) == 0 || len(config.GidMappings) == 0 {
+		return fmt.Errorf("invalid mount %+v: not yet supported to use GidMappings/UidMappings in a mount without also using a user namespace", m)
```
Let's unify the discussion in one thread, then change all the relevant code can be done easily :)

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -253,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+		return fmt.Errorf("invalid mount %+v: not yet supported to use GidMappings/UidMappings in a mount without also using a user namespace", m)
+	}
+	if !sameMapping(config.UidMappings, m.UIDMappings) {
+		return fmt.Errorf("invalid mount %+v: not yet supported for the mount uidMappings to be different than user namespace uidMapping", m)
```
Updated, thanks!

rata commented on libcontainer/configs/validate/validator.go:
```diff
@@ -253,16 +253,72 @@ func cgroupsCheck(config *configs.Config) error {
+	if config.RootlessEUID {
+		return fmt.Errorf("invalid mount %+v: GidMappings/UidMappings is not supported when runc is being launched with EUID != 0, needs CAP_SYS_ADMIN on the runc parent's user namespace", m)
+	}
+	if len(config.UidMappings) == 0 || len(config.GidMappings) == 0 {
```
I agree another PR will be better. Wanna create an issue for that? :)

rata replied:
Created an issue for this: https://github.com/opencontainers/runc/issues/3936  :)

rata commented:
@thaJeztah Thanks a lot for the detailed review! I fixed most of the comments and amended to the proper commits, you can see the overall diff here: https://github.com/opencontainers/runc/compare/f4d70ef6b968f2b9a39ef07396048557563d3fe1..b6d4681d6ee0a0d0da1b6f8050287ff094cf934b, it is just trivial.

I hope this LGTY now :-D

rata commented:
The "ci / cross-i386 (pull_request) " test is failing and is unrelated to this PR. Not sure what caused the broken packages, maybe github but maybe some packages changed?

thaJeztah commented on libcontainer/container_linux.go:
```diff
@@ -541,6 +541,43 @@ func (c *Container) shouldSendMountSources() bool {
+
+		if !condition(m) {
+			continue
+		}
```
Yes, I'm probably fine with keeping `-1` as value. This was mostly me pondering "do we need a special value, or would default (`0`) still work?

> Wanna create an issue?

done! https://github.com/opencontainers/runc/issues/3934

rata replied:
Thanks!

thaJeztah commented:
We've had some issues with packages in the containerd repository (which was related to mirrors configured on the workers that only have a subset of architectures).

I gave CI a kick (hoping it was just a fluke) 🤞

thaJeztah reviewed (approved):
LGTM, thanks so much!

⚠️ ⚠️ I did not review the C code extensively; I'm not a C-coder, and don't want to pretend being one, so other than "a proper glance", haven't reviewed those commits. Reviews from others on those files therefore would be appreciated 👍

lifubang commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -882,6 +982,18 @@ void nsexec(void)
+					send_idmapsources(syncfd, stage1_pid, config.idmapsources,
+							  config.idmapsources_len);
+					s = SYNC_MOUNT_IDMAP_ACK;
+					write_log(DEBUG, "idmap sources done, sending SYNC_MOUNT_IDMAP_ACK");
```
@rata A small suggestion: I think you used this log when you were debugging. If you want to remain this debug log, please add `stage-xxx` in the head. But I suggest to remove this log to keep the same log as `mount sources`.

rata replied:
Sure, I deleted this one. Thanks for checking the consistency!

When we wrote the mount sources patch with a co-worker we didn't add any debug lines, but when writing this PR I did, so I created a PR to add them for mount sources too: https://github.com/opencontainers/runc/pull/3734/files (I wrote that PR before finishing this one, probably)

I'm on the fence if these lines are useful or not, but if they are, I'll open a new PR adding them to the idmap and mount sources cases.

lifubang commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -1053,6 +1165,22 @@ void nsexec(void)
+
+				/* Receive and install all idmap fds. */
+				receive_idmapsources(syncfd);
+				write_log(DEBUG, "received opened idmap sources");
```
@rata A small suggestion: I think you used this log when you were debugging. If you want to remain this debug log, please add `stage-xxx` in the head. But I suggest to remove this log to keep the same log as `mount sources`.

rata replied:
Deleted this one too, thanks!

lifubang reviewed (commented):
Two small suggestions. we aleady have 3 LGTMs, I think we can move forward now.

rata commented:
@lifubang thanks, fixed that too. Feel free to merge now! :tada:

lifubang commented:
Thanks everyone working on this.

cyphar reviewed (commented):
Ah, this was merged while I was reviewing it. I guess I'll send a fix-up PR then...

rata commented on libcontainer/nsenter/nsexec.c:
```diff
@@ -641,6 +664,83 @@ void try_unshare(int flags, const char *msg)
+			continue;
+		}
+
+		int fd_tree = sys_open_tree(-EBADF, idmap_src,
```
We are enforcing that the path is a abs dir. If we use `AT_FDCWD` it will work (if the validation is skipped for some reason), while with this IIRC it will fail if the path is a rel path.

Why is this suggestion better?

rata commented:
@cyphar Sure, feel free to open a PR and tag me if that works for you. Otherwise, let me know what suggested changes you had in mind

lifubang commented:
> Ah, this was merged while I was reviewing it.

Sorry to interupt your code review. Maybe you can set this PR's label to `status/2-codereview` if the label is `status/4-merge`(This is learned from @thaJeztah ^_^) when you are doing an important code review. Thanks.

By the way, I think all the maintainers should have a msg group in a instant messaging tool If convenient. Then we can announce some important things in this msg group.

lifubang commented:
> > msg
> 
> Do you have an account in `opencontainers.slack.com` ? If you don't have, you can create one in https://communityinviter.com/apps/opencontainers/join-the-oci-community

Has created one account: Lifubang. Can you bring me in `runc` user group if there is? Thanks.

AkihiroSuda commented:
Done, but not much active currently

lifubang commented on libcontainer/rootfs_linux.go:
```diff
@@ -81,11 +86,20 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds []int) (err
 	for i, m := range config.Mounts {
-		entry := mountEntry{Mount: m}
+		entry := mountEntry{Mount: m, idmapFD: -1}
 		// Just before the loop we checked that if not empty, len(mountFds) == len(config.Mounts).
```
This is from #3717, it is not `len(mountFds)`, but `len(mountFds.sourceFds)`. I think line 91-92 could be removed.

lifubang commented on libcontainer/rootfs_linux.go:
```diff
@@ -81,11 +86,20 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds []int) (err
-		entry := mountEntry{Mount: m}
+		entry := mountEntry{Mount: m, idmapFD: -1}
 		// Just before the loop we checked that if not empty, len(mountFds) == len(config.Mounts).
 		// Therefore, we can access mountFds[i] without any concerns.
```
There is no `mountFds[i]`.

lifubang commented on libcontainer/rootfs_linux.go:
```diff
@@ -81,11 +86,20 @@ func prepareRootfs(pipe io.ReadWriter, iConfig *initConfig, mountFds []int) (err
+			entry.srcFD = "/proc/self/fd/" + strconv.Itoa(mountFds.sourceFds[i])
+		}
+
+		// We validated before we can access idmapFds[i].
```
Please remove this comment too.
